File: python/revo1/revo1_basic_collector.py
"""
Revo1 Basic Version - High-Performance Data Collector Example with Trajectory Control

This example demonstrates high-performance motor data collection using shared buffers
while simultaneously running trajectory control:
- Motor status data collection at 100Hz
- Trajectory control at 100Hz
- Efficient data collection without GIL overhead
- Observe motor response to trajectory commands in real-time

Hardware Support:
- ✅ Revo1 Basic
- ✅ Revo1 Touch

Usage:
    python revo1_basic_collector.py
"""

import asyncio
import sys
import time
from revo1_utils import *
from utils import setup_shutdown_event

# Trajectory control parameters
TRAJ_LEN = 20           # Number of trajectory points
CTRL_INTERVAL = 0.01    # Control interval: 10ms (100Hz)
ENABLE_CONTROL = True   # Enable/disable trajectory control


async def main():
    """Main function: High-performance motor data collection"""

    # Setup shutdown event listener
    shutdown_event = setup_shutdown_event(logger)

    # Open Modbus connection
    logger.info("Connecting to device...")
    (client, detected_slave_id) = await open_modbus_revo1()
    slave_id = detected_slave_id

    # Get device info
    device_info = await client.get_device_info(slave_id)
    logger.info(f"Device info: {device_info.description}")

    # ========================================================================
    # Create shared buffer
    # ========================================================================

    logger.info("Creating shared buffer...")
    motor_buffer = libstark.MotorStatusBuffer(max_size=1000)
    logger.info("Shared buffer created")

    # ========================================================================
    # Create and start data collector - Basic mode
    # ========================================================================

    logger.info("Creating data collector (Basic mode)...")

    # Collection frequency (platform-dependent)
    # Linux: 100Hz (high-performance serial port)
    # Windows/macOS: 20Hz (conservative for compatibility)
    import platform
    is_linux = platform.system() == 'Linux'
    motor_frequency = 100 if is_linux else 20
    
    logger.info(f"Platform: {platform.system()}, Motor: {motor_frequency}Hz")

    collector = libstark.DataCollector.new_basic(
        ctx=client,
        motor_buffer=motor_buffer,
        slave_id=slave_id,
        motor_frequency=motor_frequency,
        enable_stats=True,
    )

    logger.info("Data collector created")

    # ========================================================================
    # Start data collection
    # ========================================================================

    logger.info("Starting data collector...")
    collector.start()
    logger.info("Data collector started (background thread)")

    # ========================================================================
    # Start trajectory control task (optional)
    # ========================================================================

    stop_flag = {'value': False}
    index_ref = {'value': 0}
    control_task = None

    if ENABLE_CONTROL:
        logger.info("Initializing trajectory control...")
        # Generate trajectory (0-100% range)
        # Note: Thumb trajectory will be auto-clamped to max 50% if needed
        trajectory = init_cosine_trajectory(traj_len=TRAJ_LEN)
        finger_id = libstark.FingerId.Ring  # Control ring finger

        control_task = asyncio.create_task(
            trajectory_control_task(
                client, slave_id, finger_id, trajectory,
                ctrl_interval=CTRL_INTERVAL,
                stop_flag_ref=stop_flag,
                index_ref=index_ref
            )
        )
        logger.info("Trajectory control task started")

    # ========================================================================
    # Start data processing task
    # ========================================================================

    # Start data processing task
    asyncio.create_task(process_motor_data(motor_buffer, index_ref if ENABLE_CONTROL else None))

    logger.info("Data processing task started")
    logger.info("Press Ctrl+C to stop...")

    # Wait for shutdown event
    await shutdown_event.wait()

    # ========================================================================
    # Stop and cleanup
    # ========================================================================

    logger.info("Stopping trajectory control...")
    if ENABLE_CONTROL and control_task:
        stop_flag['value'] = True
        await control_task
        logger.info("Trajectory control stopped")

    logger.info("Stopping data collector...")
    collector.stop()
    # Give a small delay to ensure logs are flushed
    await asyncio.sleep(0.1)
    logger.info("Data collector stopped")

    # Cleanup resources
    libstark.modbus_close(client)
    logger.info("Modbus client closed")
    sys.exit(0)


async def process_motor_data(motor_buffer, index_ref=None):
    """
    Process motor data
    
    Args:
        motor_buffer: Motor status buffer
        index_ref: Optional trajectory index reference for correlation
    """
    logger.info("Motor data processing task started")
    process_count = 0

    while True:
        try:
            if motor_buffer.is_empty():
                await asyncio.sleep(0.01)
                continue

            # Batch get data
            data_batch = motor_buffer.pop_all()

            if len(data_batch) > 0:
                if process_count % 2 == 0:  # Print every 2 times
                    traj_info = f" [Traj Index: {index_ref['value']}]" if index_ref else ""
                    logger.debug(f"[{process_count}]{traj_info} Processed {len(data_batch)} motor data")

                    # Print latest data
                    latest = data_batch[-1]
                    # logger.info(f"Positions: {latest.positions[:6]}")
                    # logger.info(f"Speeds: {latest.speeds[:6]}")
                    # logger.info(f"Currents: {latest.currents[:6]}")
                    logger.info(f"Finger Detail: {latest.description}")

                process_count += 1

            await asyncio.sleep(0.01)  # 10ms processing interval

        except Exception as e:
            logger.error(f"Motor data processing error: {e}")
            await asyncio.sleep(1)


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        logger.info("User interrupted")
        sys.exit(0)
    except Exception as e:
        logger.error(f"Error: {e}", exc_info=True)
        sys.exit(1)

File: linux/revo1/revo1_basic_collector.cpp
#include "stark-sdk.h"
#include "../common/stark_common.h"
#include "../common/trajectory_control.h"
#include <stdio.h>
#include <unistd.h>
#include <signal.h>

/**
 * Revo1 Basic Version - High-Performance Data Collector Example with Trajectory Control
 * 
 * This example demonstrates high-performance motor data collection using shared buffers
 * while simultaneously running trajectory control:
 * - Motor status data collection at 50Hz (Linux) / 20Hz (other platforms)
 * - Trajectory control at 100Hz
 * - Efficient data transfer via shared buffer API
 * - Lock-free buffer design for minimal overhead
 * - Observe motor response to trajectory commands in real-time
 * 
 * Hardware Support:
 * - ✅ Revo1 Basic (基础版)
 * - ✅ Revo1 Touch (电容式触觉)
 * 
 * Compile:
 *   make revo1_basic_collector
 * 
 * Run:
 *   ./revo1_basic_collector
 */

// Trajectory control parameters
#define TRAJ_LEN 20           // Number of trajectory points
#define CTRL_FREQUENCY 100    // Control frequency: 100Hz
#define ENABLE_CONTROL 1      // Enable/disable trajectory control (1=enable, 0=disable)

// Global variable for signal handling
static volatile int keep_running = 1;

void signal_handler(int signum) {
  printf("\n[INFO] Received signal %d, stopping...\n", signum);
  keep_running = 0;
}

int main(int argc, char const *argv[]) {
  // Setup signal handlers
  setup_signal_handlers();
  signal(SIGINT, signal_handler);
  signal(SIGTERM, signal_handler);

  // Initialize SDK
  init_cfg(STARK_PROTOCOL_TYPE_MODBUS, LOG_LEVEL_INFO);
  
  // Auto-detect device
  auto cfg = auto_detect_modbus_revo1("/dev/ttyUSB0", true);
  if (cfg == NULL) {
    fprintf(stderr, "[ERROR] Failed to auto-detect Modbus device.\n");
    return -1;
  }
  
  uint8_t slave_id = cfg->slave_id;
  printf("[INFO] Detected device on %s, slave_id=%d, baudrate=%d\n", 
         cfg->port_name, slave_id, cfg->baudrate);
  
  // Open device
  auto handle = modbus_open(cfg->port_name, cfg->baudrate);
  free_device_config(cfg);
  
  if (handle == NULL) {
    fprintf(stderr, "[ERROR] Failed to open device.\n");
    return -1;
  }

  // Get device info
  auto info = stark_get_device_info(handle, slave_id);
  if (info != NULL) {
    printf("[INFO] Device: SKU=%d, SN=%s, FW=%s\n",
           (uint8_t)info->sku_type, info->serial_number, info->firmware_version);
    free_device_info(info);
  }

  // ========================================================================
  // Create shared buffer
  // ========================================================================
  
  printf("[INFO] Creating shared buffer...\n");
  
  // Motor status buffer (max 1000 items)
  auto motor_buffer = motor_buffer_new(1000);
  if (motor_buffer == NULL) {
    fprintf(stderr, "[ERROR] Failed to create motor buffer.\n");
    return -1;
  }

  // ========================================================================
  // Create and start data collector - Basic mode
  // ========================================================================
  
  printf("[INFO] Creating data collector (Basic mode)...\n");
  
  // Platform-dependent frequencies
  // Linux: 50Hz (Revo1 is slower than Revo2)
  // Other platforms: 20Hz (conservative for compatibility)
  #ifdef __linux__
    const uint32_t motor_frequency = 50;   // Linux: 50Hz (Revo1)
  #else
    const uint32_t motor_frequency = 20;   // Windows/macOS: 20Hz
  #endif
  
  printf("[INFO] Motor frequency: %dHz\n", motor_frequency);
  
  auto collector = data_collector_new_basic(
    handle,                    // Device context
    motor_buffer,              // Motor buffer
    slave_id,                  // Slave ID
    motor_frequency,           // Motor frequency (platform-dependent)
    1                          // Enable stats
  );
  
  if (collector == NULL) {
    fprintf(stderr, "[ERROR] Failed to create data collector.\n");
    motor_buffer_free(motor_buffer);
    return -1;
  }

  // Start data collection
  printf("[INFO] Starting data collector...\n");
  if (data_collector_start(collector) != 0) {
    fprintf(stderr, "[ERROR] Failed to start data collector.\n");
    data_collector_free(collector);
    motor_buffer_free(motor_buffer);
    return -1;
  }

  printf("[INFO] Data collector started successfully!\n");
  printf("[INFO] Basic mode: Motor data collection at %dHz\n", motor_frequency);

  // ========================================================================
  // Start trajectory control (optional)
  // ========================================================================
  
  TrajectoryControl traj_ctrl;
  uint16_t *trajectory = NULL;
  
  #if ENABLE_CONTROL
    printf("[INFO] Initializing trajectory control...\n");
    
    // Generate cosine trajectory (0-1000 representing 0-100%)
    // Note: Thumb trajectory will be auto-clamped to max 500 if needed
    trajectory = init_cosine_trajectory(TRAJ_LEN, 0, 1000);
    if (trajectory == NULL) {
      fprintf(stderr, "[ERROR] Failed to create trajectory.\n");
      data_collector_free(collector);
      motor_buffer_free(motor_buffer);
      return -1;
    }
    
    // Initialize trajectory control for ring finger
    trajectory_control_init(
      &traj_ctrl,
      handle,
      slave_id,
      STARK_FINGER_ID_RING,  // Control ring finger
      trajectory,
      TRAJ_LEN,
      CTRL_FREQUENCY
    );
    
    // Start trajectory control
    if (trajectory_control_start(&traj_ctrl) != 0) {
      fprintf(stderr, "[ERROR] Failed to start trajectory control.\n");
      free(trajectory);
      data_collector_free(collector);
      motor_buffer_free(motor_buffer);
      return -1;
    }
    
    printf("[INFO] Trajectory control started (Ring finger, %dHz)\n", CTRL_FREQUENCY);
  #endif
  
  printf("[INFO] Press Ctrl+C to stop...\n\n");

  // ========================================================================
  // Main loop: Read data from buffer
  // ========================================================================
  
  // Allocate array for reading data
  const size_t MAX_MOTOR_DATA = 1000;
  MotorStatusData motor_data[MAX_MOTOR_DATA];
  size_t motor_count = 0;
  
  int loop_count = 0;
  
  while (keep_running) {
    usleep(100 * 1000); // Sleep 100ms
    loop_count++;
    
    // Read motor status every loop
    size_t buffer_len = motor_buffer_len(motor_buffer);
    if (buffer_len > 0 && loop_count % 2 == 0) {  // Print every 2 times
      motor_buffer_pop_all(motor_buffer, motor_data, &motor_count);
      
      if (motor_count > 0) {
        auto latest = &motor_data[motor_count - 1];
        
        #if ENABLE_CONTROL
          size_t traj_index = trajectory_control_get_index(&traj_ctrl);
          printf("[Motor] Count=%zu [Traj Index: %zu], Latest: Pos=[%hu,%hu,%hu,%hu,%hu,%hu]\n",
                 motor_count, traj_index,
                 latest->positions[0], latest->positions[1], latest->positions[2],
                 latest->positions[3], latest->positions[4], latest->positions[5]);
        #else
          printf("[Motor] Count=%zu, Latest: Pos=[%hu,%hu,%hu,%hu,%hu,%hu]\n",
                 motor_count,
                 latest->positions[0], latest->positions[1], latest->positions[2],
                 latest->positions[3], latest->positions[4], latest->positions[5]);
        #endif
        
        printf("        Speeds=[%hd,%hd,%hd,%hd,%hd,%hd]\n",
               latest->speeds[0], latest->speeds[1], latest->speeds[2],
               latest->speeds[3], latest->speeds[4], latest->speeds[5]);
        printf("        Currents=[%hd,%hd,%hd,%hd,%hd,%hd]\n\n",
               latest->currents[0], latest->currents[1], latest->currents[2],
               latest->currents[3], latest->currents[4], latest->currents[5]);
      }
    }
  }

  // ========================================================================
  // Cleanup
  // ========================================================================
  
  #if ENABLE_CONTROL
    printf("\n[INFO] Stopping trajectory control...\n");
    trajectory_control_stop(&traj_ctrl);
    if (trajectory) {
      free(trajectory);
    }
  #endif
  
  printf("\n[INFO] Stopping data collector...\n");
  data_collector_stop(collector);
  
  printf("[INFO] Cleaning up...\n");
  data_collector_free(collector);
  motor_buffer_free(motor_buffer);
  
  printf("[INFO] Done!\n");
  
  return 0;
}
